src/test/app/TheoreticalQuality_test.cpp

```diff
@@ -0,0 +1,568 @@
+    {
+        testcase("Direct Step");
+
+        // clang-format off
+
+        // Set up a payment through four accounts: alice -> bob -> carol -> dan
+        // For each relavent trust line on the path, there are three things that can vary:
+        //  1) input quality
+        //  2) output quality
+        //  3) debt direction
+        // For each account, there is one thing that can vary:
+        //  1) transfer rate
+
+        // clang-format on
+
+        using namespace jtx;
+
+        auto const currency = to_currency("USD");
+
+        constexpr std::size_t const numAccounts = 4;
+
+        // There are three relavent trust lines: `alice->bob`, `bob->carol`, and
+        // `carol->dan`. There are four accounts. If we count the number of
+        // combinations of parameters where a parameter is changed from its
+        // default value, there are
+        // 2^(num_trust_lines*num_trust_qualities+numAccounts) combinations of
+        // values to test, or 2^13 combinations. Use this value to set the
+        // number of iterations. Note however that many of these parameter
+        // combinations run essentially the same test. For example, changing the
+        // quality values for bob and carol are test almost the same thing.
+        // Similarly, changing the transfer rates on bob and carol test almost
+        // the same thing. Instead of systematically running these 8k tests,
+        // randomly sample the test space.
+        int const numTestIterations = reqNumIterations.value_or(500);
+
+        constexpr std::uint32_t paymentAmount = 1;
+
+        // Class to randommly set account transfer rates, qualities, and other
+        // params.
+        RandomAccountParams rndAccParams;
+
+        // Tests are sped up by a factor of 2 if a new environment isn't created
+        // on every iteration.
+        Env env(*this, supported_amendments());
+        for (int i = 0; i < numTestIterations; ++i)
+        {
+            auto const iterAsStr = to_string(i);
+            // New set of accounts on every iteration so the environment doesn't
+            // need to be recreated (2x speedup)
+            auto const alice = Account("alice" + iterAsStr);
+            auto const bob = Account("bob" + iterAsStr);
+            auto const carol = Account("carol" + iterAsStr);
+            auto const dan = Account("dan" + iterAsStr);
+            std::array<Account, numAccounts> accounts{alice, bob, carol, dan};
+            static_assert(
+                numAccounts == 4, "Path is only correct for four accounts");
+            path const accountsPath(accounts[1], accounts[2]);
+            env.fund(XRP(10000), alice, bob, carol, dan);
+            env.close();
+
+            // iterate through all pairs of accounts, randomly set the transfer
+            // rate, qIn, qOut, and if the account issues or redeems
+            for (std::size_t i = 0; i < numAccounts; ++i)
```

Comment:
This `i` shadows the outer `i`.  Maybe `j` would be a better choice (assuming no one thinks it's a `Journal`?  No, `j` is used below...  `k` maybe?

At any rate, the shadowing really ought to be fixed.  It's easy for a reader to get confused between the inner and outer `i`.

Reply:
Fixed in 8ac0095f7 `[fold] Fix spelling errors and rename some variables`